### DISCUSSION TITLE
Headless submissions via custom REST API / PHP?

### DISCUSSION THREAD

**Discussion by mandrasch** [verbb/formie#2034] [Q&A] [answered]
Hi,

I'm researching for the following use case:

1. NextJS site which has a simple form (name, multiline text, emailaddress)
2. Submit it to CraftCMS via REST API / `fetch()` 

I know, Formie has good [GraphQL support](https://verbb.io/craft-plugins/formie/docs/feature-tour/headless) - but in the use case we would like to do it via [REST API](https://putyourlightson.com/articles/building-a-restful-api-in-craft-cms). 

I also know there are some things like [Controller Actions ](https://craftcms.com/docs/5.x/reference/controller-actions.html) - but haven't found an info about this in Formie docs.

Is there a quick way to achieve sending POST data, return validation object - or successful response when submission was made and is stored in Formie?
 
Thx in advance for any hints!

Best regards,
Matthias

**Comment by engram-design** [comment, marked as answer]
The best method is to use GraphQL mutations for headless, but using REST is also possible. Your best bet is to mimic a POST submission just like a traditional form would. We don't have docs about this, purely because it's very straightforward to create a Twig template, render your form, test submission and inspect the payload sent to the server. So long as you copy that, you're fine.

For example:

```js
// Method 1: construct an object of data
var body = {
"action": "formie/submissions/submit",
"handle": "contactForm",
"siteId": "1",
"fields[yourName]": "dave",
"fields[emailAddress]": "test@test.test",
"fields[message]": "this is a message",
};

// Append CSRF token
window.csrfTokenName = "{{ craft.app.config.general.csrfTokenName|e('js') }}";
window.csrfTokenValue = "{{ craft.app.request.csrfToken|e('js') }}";
body[window.csrfTokenName] = window.csrfTokenValue;

fetch('/', {
    method: 'post',
    body: body,
    headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
        "X-Requested-With": "XMLHttpRequest",
    },
}).then(function(response) {
    response.json().then(function(data) {
        console.log(data);
    });
})

// Method 2: construct from a form, using FormData
let formElem = document.getElementById('form');
let data = new FormData(formElem);

fetch('/', {
    method: 'post',
    body: data,
    headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
        "X-Requested-With": "XMLHttpRequest",
    },
}).then(function(response) {
    response.json().then(function(data) {
        console.log(data);
    });
})
```

But because you're doing all the submitting on your own, you're going to need to handle things like captchas on your own, such as reCAPTCHA and append those to the payload.

**Reply by mandrasch** [reply]
Hi! Awesome, thanks very much for quick reply and detailed example and re-assurance! 👍 👍 Thanks for the hint regarding captchas as well.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.3](https://github.com/verbb/formie/releases/tag/4.0.0-beta.3).

We've got an entirely new API for handling headless with either REST or GraphQL now. You can also check out our [starters](https://verbb.io/craft-plugins/formie/docs/v4/frontend/frontend-assets#frameworks) that illustrate this in a few different frameworks. We'll also likely add a "vanilla JS/TS" example too.

**Reply by mandrasch** [reply]
Oh wow, great! Too bad I'm not working with craft right now, but otherwise I would have given it a try for sure! Great addition. 👏
